### PR TITLE
Feature/constraints

### PR DESCRIPTION
**Pull request by StewStrong**

### CODE REVIEW

File: api-game/src/main/kotlin/org/valkyrienskies/core/apigame/constraints/VSConstraintType.kt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package org.valkyrienskies.core.apigame.constraints
+
+/**
+ * An enum that identifies different types of constraints. Do not use [VSConstraintType] for more than one constraint!
+ */
+enum class VSConstraintType {
```
Comment (ewoudje) [inline api-game/src/main/kotlin/org/valkyrienskies/core/apigame/constraints/VSConstraintType.kt:6]: Isn't it better to just seal the interface VSConstraint and remove the enum

File: api-game/src/main/kotlin/org/valkyrienskies/core/apigame/constraints/VSConstraintAndId.kt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package org.valkyrienskies.core.apigame.constraints
+
+/**
+ * The constraint id is stored here because it is assigned by the physics engine, not when the constraint is created.
+ */
+data class VSConstraintAndId(
```
Comment (ewoudje) [inline api-game/src/main/kotlin/org/valkyrienskies/core/apigame/constraints/VSConstraintAndId.kt:6]: Im a bit confused by this comment. Is the id assigned from krunch? bcs as far as i see it is just a counter in ServerWorld

**Review by ewoudje — COMMENTED** [review]
Looking great!

File: impl/src/main/kotlin/org/valkyrienskies/core/impl/pipelines/VSGamePipelineStage.kt [hunk 3]
```diff
@@ -208,8 +221,72 @@ class VSGamePipelineStage @Inject constructor(private val shipWorld: ShipObjectS
...
 
+        // region Convert the coordinates of the constraints to be relative to center of mass
+        val constraintsCreatedThisTick: MutableList<VSConstraintAndId> = ArrayList()
+        lastTickChanges.constraintsCreatedThisTick.forEach {
+            if (it.vsConstraint !is VSForceConstraint) {
+                constraintsCreatedThisTick.add(it)
+            } else {
+                val adjusted = adjustConstraint(it.vsConstraint)
+                if (adjusted == null) {
+                    println("Failed to adjust a constraint. Was a ship deleted?")
```
Comment (ewoudje) [inline impl/src/main/kotlin/org/valkyrienskies/core/impl/pipelines/VSGamePipelineStage.kt:232]: Logger!!

File: api-game/src/main/kotlin/org/valkyrienskies/core/apigame/constraints/VSConstraintAndId.kt (hunk 2 above)
Comment (StewStrong) [reply api-game/src/main/kotlin/org/valkyrienskies/core/apigame/constraints/VSConstraintAndId.kt:3]: Yeah that comment was incorrect. I deleted it in https://github.com/ValkyrienSkies/vs-core/pull/44/commits/172469a6952a7958be880ec1f765d6454b52a827

File: api-game/src/main/kotlin/org/valkyrienskies/core/apigame/constraints/VSConstraintType.kt (hunk 1 above)
Comment (StewStrong) [reply api-game/src/main/kotlin/org/valkyrienskies/core/apigame/constraints/VSConstraintType.kt:6]: Sealed in https://github.com/ValkyrienSkies/vs-core/pull/44/commits/e83b04ddbb8ddf21a59872cf10d3c8f76b4c20d5

**Review by ewoudje — APPROVED** [review]
Seems good to me.

**Pull request merged**